api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -175,9 +183,11 @@ private String escape(String string) {
   public String partitionToPath(StructLike data) {
     StringBuilder sb = new StringBuilder();
     Class<?>[] javaClasses = javaClasses();
+    List<Types.NestedField> outputFields = partitionType().fields();
     for (int i = 0; i < javaClasses.length; i += 1) {
       PartitionField field = fields[i];
-      String valueString = field.transform().toHumanString(get(data, i, javaClasses[i]));
+      String valueString =
+          field.transform().toHumanString(outputFields.get(i).type(), get(data, i, javaClasses[i]));
```

Comment:
nit: If you define a helper var for `outputFields.get(i).type()`, this would fit on one line.

Reply:
Fixed.